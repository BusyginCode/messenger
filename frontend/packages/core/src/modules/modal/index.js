export * from './modal.actions'
export * from './modal.selectors'

export { default } from './modal.reducer'
