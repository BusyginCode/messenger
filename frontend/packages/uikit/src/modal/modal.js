import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import ReactModal from 'react-modal'
import cn from 'classnames'
import snakeCase from 'lodash/snakeCase'
import CloseIcon from '@messenger/icons/lib/icon/svgr/m/close'
import Box from '../box'
import Text from '../text'
import styles from './modal.scss'

class Modal extends PureComponent {
  static propTypes = {
    children: PropTypes.node,
    header: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
    name: PropTypes.string,
    className: PropTypes.string,
    classNameOverlay: PropTypes.string,
    classNameContent: PropTypes.string,
    isOpened: PropTypes.bool,
    close: PropTypes.func,
    overflowContent: PropTypes.node,
    isCustom: PropTypes.bool,
    isCustomHeader: PropTypes.bool,
    isCanClose: PropTypes.bool,
    shouldCloseOnOverlayClick: PropTypes.bool,
    shouldCloseOnEsc: PropTypes.bool
  }

  static defaultProps = {
    isCanClose: true,
    shouldCloseOnOverlayClick: true,
    shouldCloseOnEsc: true
  }

  state = {
    overlayRef: null
  }

  overlayRef = ref => this.setState({ overlayRef: ref })

  getOverlayRef = () => this.state.overlayRef

  renderHeader = (header, isCustom) => {
    if (isCustom) {
      return header
    }

    return (
      <Box justify="around" align="center" inject className={styles.header}>
        <Text size="xl" transform="capitalize" semibold>
          {header}
        </Text>
      </Box>
    )
  }

  render() {
    const {
      children,
      header,
      name,
      overflowContent,
      isOpened,
      close,
      isCustom,
      isCustomHeader,
      isCanClose,
      className,
      classNameOverlay,
      classNameContent,

      shouldCloseOnOverlayClick,
      shouldCloseOnEsc
    } = this.props

    if (isOpened) {
      if (isCustom) {
        return (
          <ReactModal
            className={cn('custom', className)}
            isOpen
            overlayClassName={classNameOverlay}
            onRequestClose={close}
            ariaHideApp={false}
            contentLabel=""
          >
            {children}
          </ReactModal>
        )
      }

      return (
        <ReactModal
          className={cn(styles.general, className)}
          isOpen
          overlayClassName={cn(classNameOverlay, styles.overlay)}
          onRequestClose={close}
          ariaHideApp={false}
          overlayRef={this.overlayRef}
          contentLabel=""
          shouldCloseOnOverlayClick={shouldCloseOnOverlayClick}
          shouldCloseOnEsc={shouldCloseOnEsc}
        >
          <div data-role={`modal-${snakeCase(name)}`} className={styles.modal}>
            {header &&
              (isCustomHeader ? (
                header
              ) : (
                <Box justify="around" align="center" inject className={styles.header}>
                  <Text data-role={`modal-${snakeCase(name)}-header`} size="xl" transform="capitalize" semibold>
                    {header}
                  </Text>
                </Box>
              ))}
            <div data-role={`modal-content-${snakeCase(name)}`} className={cn(styles.content, classNameContent)}>
              {React.cloneElement(children, {
                renderHeader: this.renderHeader,
                getOverlayRef: this.getOverlayRef
              })}
            </div>
          </div>
          {isCanClose && <CloseIcon data-role="modal-close-icon" className={styles.close} onClick={close} />}
          {overflowContent}
        </ReactModal>
      )
    }

    return null
  }
}

export default Modal
