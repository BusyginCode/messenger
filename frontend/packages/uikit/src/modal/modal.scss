.general {
  position: relative;
  margin: auto;
}

.modal {
  position: relative;
  box-sizing: border-box;
  box-shadow: var(--b-shadow-xxl);
  border-radius: var(--b-border-radius-m);
  overflow: hidden;

  .header {
    position: relative;
    height: 64px;
    padding: var(--b-spacing-m) var(--b-spacing-l);
    line-height: var(--b-line-height-xxl);
    background-color: var(--b-color-body8_drk);
    color: var(--b-color-body6);
  }
}

.content {
  background-color: var(--b-color-body6);
}

.overlay {
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background-color: rgba(49, 69, 89, 0.8);
  display: flex;
  align-items: center;
  justify-content: center;
  z-index: 12;
  overflow-y: scroll;
  padding: var(--b-spacing-xl) 0 var(--b-spacing-xxl) 0;
}

.close {
  width: var(--b-size-xs);
  height: var(--b-size-xs);
  position: absolute;
  right: calc(-1 * (var(--b-size-xs) + var(--b-spacing-s)));
  top: calc(-1 * (var(--b-size-xs) + var(--b-spacing-s)));
  color: var(--b-color-body3);
  cursor: pointer;
}

:global {
  .ReactModal__Body--open {
    overflow: hidden;
  }
}
