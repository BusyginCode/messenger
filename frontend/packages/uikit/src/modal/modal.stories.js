import React from 'react'
import { storiesOf } from '@storybook/react'
import Modal from './modal'
import Button from '../button'

storiesOf('RFC/UIKit/Modal', module)
  .add('custom', () => (
    <Modal isOpened>
      <div style={{ backgroundColor: 'red', width: 300, height: 300 }}>Red box</div>
    </Modal>
  ))
  .add('with header controls', () => (
    <Modal
      isOpened
      leftHeaderControl={<Button>Whatever</Button>}
      rightHeaderControl={<Button>Whatever</Button>}
      header={'Something'}
    >
      <div style={{ width: 400, height: 500 }}>Something</div>
    </Modal>
  ))
