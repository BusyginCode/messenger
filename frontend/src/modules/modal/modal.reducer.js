import { handleActions } from 'redux-actions'
import omit from 'lodash/omit'
import { openModal, closeModal } from './modal.actions'

const initialState = {}

const handlerOpen = (state, { payload: { name = `default_${Math.random()}`, ...rest } }) => ({
  ...state,
  [name]: {
    ...rest,
    name
  }
})

const handlerClose = (state, { payload: name }) => omit(state, [name])

const modalReducer = handleActions(
  {
    [openModal]: handlerOpen,
    [closeModal]: handlerClose
  },
  initialState
)

export default modalReducer
