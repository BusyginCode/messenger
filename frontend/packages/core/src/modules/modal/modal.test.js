import { openModal, closeModal } from './modal.actions'
import { getModal } from './modal.selectors'
import reducer from './modal.reducer'

describe('modal actions', () => {
  it('should open correctly', () => {
    expect(openModal({ name: 'MODAL', props: {}, modalProps: {} })).toMatchSnapshot()
  })

  it('should close correctly', () => {
    expect(closeModal('MODAL')).toMatchSnapshot()
  })
})

describe('auth selectors', () => {
  it('should getNotifications correctly', () => {
    const modal = {}
    expect(getModal({ modal })).toBe(modal)
  })
})

describe('auth reducer', () => {
  it('should reduce initialState correctly', () => {
    expect(reducer(undefined, {})).toMatchSnapshot()
  })

  it('should reduce open correctly', () => {
    expect(reducer({}, openModal({ name: 'MODAL', props: {} }))).toMatchSnapshot()
  })

  it('should reduce close correctly', () => {
    expect(reducer({ MODAL: {} }, closeModal('MODAL'))).toMatchSnapshot()
  })
})
