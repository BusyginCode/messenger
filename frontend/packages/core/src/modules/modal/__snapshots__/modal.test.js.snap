// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`auth reducer should reduce close correctly 1`] = `Object {}`;

exports[`auth reducer should reduce initialState correctly 1`] = `Object {}`;

exports[`auth reducer should reduce open correctly 1`] = `
Object {
  "MODAL": Object {
    "name": "MODAL",
    "props": Object {},
  },
}
`;

exports[`modal actions should close correctly 1`] = `
Object {
  "payload": "MODAL",
  "type": "MODAL_CLOSE",
}
`;

exports[`modal actions should open correctly 1`] = `
Object {
  "payload": Object {
    "modalProps": Object {},
    "name": "MODAL",
    "props": Object {},
  },
  "type": "MODAL_OPEN",
}
`;
